samples/test/data_passing.py

```diff
@@ -0,0 +1,151 @@
+# 5. Pass task output that gets consumed as value. (Common)
+# 6. Pass task output that gets consumed as file. (Common)
+
+# The only restriction on types is that when both upstream output and downstream input have types, the types must match.
+
+# %%
+import kfp
+from kfp.components import create_component_from_func, InputPath, OutputPath
+
+
+# Components
+# Produce
+@create_component_from_func
+def produce_anything(data_path: OutputPath()):
+    with open(data_path, "w") as f:
+        f.write("produce_anything")
+
+
+@create_component_from_func
+def produce_something(data_path: OutputPath("Something")):
+    with open(data_path, "w") as f:
+        f.write("produce_something")
+
+
+@create_component_from_func
+def produce_something2() -> 'Something':
+    return "produce_something2"
+
+
+@create_component_from_func
+def produce_string() -> str:
+    return "produce_string"
+
+
+# Consume as value
+@create_component_from_func
+def consume_anything_as_value(data):
+    print("consume_anything_as_value: " + data)
+
+
+@create_component_from_func
+def consume_something_as_value(data: "Something"):
+    print("consume_something_as_value: " + data)
+
+
+@create_component_from_func
+def consume_string_as_value(data: str):
+    print("consume_string_as_value: " + data)
+
+
+# Consume as file
+@create_component_from_func
+def consume_anything_as_file(data_path: InputPath()):
+    with open(data_path) as f:
+        print("consume_anything_as_file: " + f.read())
+
+
+@create_component_from_func
+def consume_something_as_file(data_path: InputPath('Something')):
+    with open(data_path) as f:
+        print("consume_something_as_file: " + f.read())
+
+
+@create_component_from_func
+def consume_string_as_file(data_path: InputPath(str)):
+    with open(data_path) as f:
+        print("consume_string_as_file: " + f.read())
+
+
+# Pipeline
+@kfp.dsl.pipeline(name='data_passing_pipeline')
+def data_passing_pipeline(
+    anything_param="anything_param",
+    something_param: "Something" = "something_param",
+    string_param: str = "string_param",
+):
+    produced_anything = produce_anything().output
+    produced_something = produce_something().output
+    produced_string = produce_string().output
+
+    # Pass constant value; consume as value
+    consume_anything_as_value("constant")
+    consume_something_as_value("constant")
+    consume_string_as_value("constant")
+
+    # Pass constant value; consume as file
+    consume_anything_as_file("constant")
+    consume_something_as_file("constant")
+    consume_string_as_file("constant")
+
+    # Pass pipeline parameter; consume as value
+    consume_anything_as_value(anything_param)
+    consume_anything_as_value(something_param)
+    consume_anything_as_value(string_param)
+    consume_something_as_value(anything_param)
+    consume_something_as_value(something_param)
+    consume_string_as_value(anything_param)
+    consume_string_as_value(string_param)
+
+    # Pass pipeline parameter; consume as file
```

Comment:
Shall we also adapt a pipeline revision that shows exactly which works in v2?

Reply:
That's the next work. Check what's broken. Check which low hanging fruits can be fixed.